donut/modules/marketplace/helpers.py

```diff
@@ -58,44 +54,259 @@ def render_with_top_marketplace_bar(template_url, **kwargs):
     else:
         num_cols = 5
 
-    cats2d = [[]]
-    width = ""
-    if num_cols != 0:
-        # if there's nothing in categories, just return default values for cats2d and width
-        num_rows = ceil(num_cats / num_cols)
+    # If there's nothing in categories, 404; something's borked.
+    if num_cols == 0:
+        return flask.render_template('404.html'), 404
+
+    num_rows = ceil(num_cats / num_cols)
+
+    # Break categories into a 2d array so that it's easy to arrange in rows
+    # in the html file.
+    cats2d = [[]] * num_rows
+    for cat_index in range(len(categories)):
+        cats2d[cat_index // num_cols].append(categories[cat_index])
+
+    # Pass the 2d category array, urls array, and width string, along with the
+    # arguments passed in to this function, on to Flask in order to render the
+    # top bar and the rest of the content.
+    return flask.render_template(
+        template_url, cats=cats2d, width=str(100.0 / num_cols), **kwargs)
+
+
+###############
+# MANAGE PAGE #
+###############
+def manage_set_active_status(item, is_active):
+    """
+    Checks permissions and then sets the is_active status of item <item> to
+    <is_active>.
+    """
+    current_user_id = get_user_id(flask.session['username'])
+    if current_user_id != get_user_id_of_item(item) and not check_permission(
+            Permissions.ADMIN):
+        return False
+
+    s = 'UPDATE marketplace_items SET item_active=%s WHERE item_id=%s'
+    with flask.g.pymysql_db.cursor() as cursor:
+        cursor.execute(s, (is_active, item))
+
+    return True
+
+
+def manage_delete_item(item_id):
+    """
+    Checks permissions and then deletes the item.
+    """
+    current_user_id = get_user_id(flask.session['username'])
+    if current_user_id != get_user_id_of_item(
+            item_id) and not check_permission(Permissions.ADMIN):
```

Comment:
I think it's right -- if the person trying to make the change is neither the owner of the item nor an admin, don't make it.  I'll agree that it's a little confusing; do you think `if not (current_user_id == ... or check_permission(...))` would be better?

Reply:
Yeah, I definitely misread this the first time I saw it. I figured it out later but guess I forgot to remove this comment. I think `if not (... or ...):` is a little more readable since it suggests you are doing something only in the exceptional case that the user is *not* allowed to do this. But totally your call.